materialize-databricks/sqlgen.go

```diff
@@ -0,0 +1,271 @@
+package main
+
+import (
+	"slices"
+	"strings"
+	"text/template"
+
+	sql "github.com/estuary/connectors/materialize-sql"
+)
+
+// databricksDialect returns a representation of the Databricks SQL dialect.
+// https://docs.databricks.com/en/sql/language-manual/index.html
+var databricksDialect = func() sql.Dialect {
+  // Although databricks does support ARRAY and MAP types, they are statically
+  // typed and the MAP type is not comparable.
+  // Databricks supports JSON extraction using the : operator, this seems like
+  // a simpler method for persisting JSON values:
+  // https://docs.databricks.com/en/sql/language-manual/sql-ref-json-path-expression.html
+	var jsonMapper = sql.NewStaticMapper("STRING", sql.WithElementConverter(sql.JsonBytesConverter))
+
+  // https://docs.databricks.com/en/sql/language-manual/sql-ref-datatypes.html
+	var typeMappings = sql.ProjectionTypeMapper{
+		sql.ARRAY:    jsonMapper,
+		sql.BINARY:   sql.NewStaticMapper("BINARY"),
+		sql.BOOLEAN:  sql.NewStaticMapper("BOOLEAN"),
+		sql.INTEGER:  sql.NewStaticMapper("BIGINT"),
+		sql.NUMBER:   sql.NewStaticMapper("DOUBLE"),
+		sql.OBJECT:   jsonMapper,
+		sql.MULTIPLE: jsonMapper,
+		sql.STRING: sql.StringTypeMapper{
+			Fallback: sql.NewStaticMapper("STRING"),
+			WithFormat: map[string]sql.TypeMapper{
+				"integer": sql.PrimaryKeyMapper{
+					PrimaryKey: sql.NewStaticMapper("STRING"),
+					Delegate:   sql.NewStaticMapper("BIGINT", sql.WithElementConverter(sql.StdStrToInt())),
+				},
+				"number": sql.PrimaryKeyMapper{
+					PrimaryKey: sql.NewStaticMapper("STRING"),
+					Delegate:   sql.NewStaticMapper("DOUBLE", sql.WithElementConverter(sql.StdStrToFloat("NaN", "Inf", "-Inf"))),
+				},
+				"date":      sql.NewStaticMapper("DATE"),
+				"date-time": sql.NewStaticMapper("TIMESTAMP"),
+			},
+		},
+	}
+	var nullable sql.TypeMapper = sql.MaybeNullableMapper{
+		NotNullText: "NOT NULL",
+		Delegate:    typeMappings,
+	}
+
+	return sql.Dialect{
+    // https://docs.databricks.com/en/sql/language-manual/sql-ref-identifiers.html
+		Identifierer: sql.IdentifierFn(sql.JoinTransform(".", 
+			sql.PassThroughTransform(
+				func(s string) bool {
+					return sql.IsSimpleIdentifier(s) && !slices.Contains(DATABRICKS_RESERVED_WORDS, strings.ToLower(s))
+				},
+        sql.QuoteTransform("`", "``"),
+      ))),
+		Literaler: sql.LiteralFn(sql.QuoteTransform("'", "\\'")),
+		Placeholderer: sql.PlaceholderFn(func(_ int) string {
+			return "?"
+		}),
+		TypeMapper:               nullable,
+		AlwaysNullableTypeMapper: sql.AlwaysNullableMapper{Delegate: typeMappings},
+	}
+}()
+
+
+// TODO: shard key in temporary table names
+// TODO: use create table USING location instead of copying data into temporary table
+var (
+	tplAll = sql.MustParseTemplate(databricksDialect, "root", `
+{{ define "temp_name_load" -}}
+flow_temp_load_table_{{ $.ShardRange }}_{{ $.Table.Binding }}
```

Comment:
@williamhbaker that's a good point, I think since we don't support having two materializations materializing to the same target table, we can just use the target table's name here and that would still give us enough isolation.